Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp

```diff
@@ -28,33 +28,55 @@
 
 #include "CookieStoreGetOptions.h"
 #include "JSDOMPromiseDeferred.h"
+#include "ScriptExecutionContext.h"
+#include "ServiceWorkerRegistration.h"
 #include <wtf/Ref.h>
+#include <wtf/URL.h>
 #include <wtf/Vector.h>
+#include <wtf/WeakPtr.h>
 
 namespace WebCore {
 
-Ref<CookieStoreManager> CookieStoreManager::create()
+Ref<CookieStoreManager> CookieStoreManager::create(ServiceWorkerRegistration& serviceWorkerRegistration)
 {
-    return adoptRef(*new CookieStoreManager);
+    return adoptRef(*new CookieStoreManager(serviceWorkerRegistration));
 }
 
-CookieStoreManager::CookieStoreManager() = default;
+CookieStoreManager::CookieStoreManager(ServiceWorkerRegistration& serviceWorkerRegistration)
+    : m_serviceWorkerRegistration(serviceWorkerRegistration)
+{
+}
 
 CookieStoreManager::~CookieStoreManager() = default;
 
-void CookieStoreManager::subscribe(Vector<CookieStoreGetOptions>&&, Ref<DeferredPromise>&& promise)
+void CookieStoreManager::subscribe(Vector<CookieStoreGetOptions>&& subscriptions, Ref<DeferredPromise>&& promise)
 {
-    promise->reject(ExceptionCode::NotSupportedError);
+    if (!m_serviceWorkerRegistration) {
+        promise->reject(Exception { ExceptionCode::InvalidStateError, "There is no service worker registration"_s });
+        return;
+    }
+
+    m_serviceWorkerRegistration->addCookieChangeSubscriptions(WTFMove(subscriptions), WTFMove(promise));
```

Comment:
We need a RefPtr to m_serviceWorkerRegistration on the stack, per smart pointer guidelines.